File: recipe-java-runtime/src/main/java/ca/derekcormier/recipe/BaseIngredientHook.java
package ca.derekcormier.recipe;

public abstract class BaseIngredientHook<T> {
    private final String ingredientName;
    private final Class<T> dataClass;

    public BaseIngredientHook(String ingredientName, Class<T> dataClass) {
        this.ingredientName = ingredientName;
        this.dataClass = dataClass;
    }

    public String getIngredientName() {
        return ingredientName;
    }

    public Class<T> getDataClass() {
        return dataClass;
    }

    public abstract void bake(T ingredient, Cake cake);
}

File: recipe-ts-runtime/src/BaseIngredientHook.ts
import { Cake } from "./Cake";

export abstract class BaseIngredientHook<T> {
    public constructor(private ingredientName: string, private snapshotClass: Function) {}

    public getIngredientName(): string {
        return this.ingredientName;
    }

    public getSnapshotClass(): Function {
        return this.snapshotClass;
    }

    public abstract bake(ingredient: T, cake: Cake): void;
}